### PR TITLE
Stack space reservation

### PR DESCRIPTION
**Pull request by paulidale**
- [x] documentation is added or updated
- [x] tests are added or updated

Add a call to allow space in a _stack_ to be reserved ahead of time.

### CODE REVIEW

File: ssl/ssl_cert.c [hunk 1]
```diff
@@ -468,7 +469,8 @@ STACK_OF(X509_NAME) *SSL_dup_CA_list(const STACK_OF(X509_NAME) *sk)
         SSLerr(SSL_F_SSL_DUP_CA_LIST, ERR_R_MALLOC_FAILURE);
         return NULL;
     }
-    for (i = 0; i < sk_X509_NAME_num(sk); i++) {
+    sk_X509_NAME_reserve(ret, num);
+    for (i = 0; i < num; i++) {
         name = X509_NAME_dup(sk_X509_NAME_value(sk, i));
         if (name == NULL || !sk_X509_NAME_push(ret, name)) {
```
Comment (richsalz) [inline ssl/ssl_cert.c]: If reserver guarantees that push can't fail for memory, why are we checking?  Is it this kind of thing that makes it a wip?

Comment (paulidale) [reply ssl/ssl_cert.c]: The code either needs to check the return from the reserve call or from each individual insertion.  It doesn't need to do both.  I took the simpler (lazier?) route by not checking the reserve return code.  If the reserve fails, it is still possible the insertions might succeed.

**Comment by paulidale** [comment]
I've addressed the same allocations as #4379 here.

**Comment by paulidale** [comment]
Travis error is unrelated.

File: include/openssl/stack.h [hunk 2]
```diff
@@ -10,6 +10,8 @@
 #ifndef HEADER_STACK_H
 # define HEADER_STACK_H
 
+# include <stddef.h>
```
Comment (levitte) [inline include/openssl/stack.h]: Why is this needed?

File: crypto/stack/stack.c [hunk 3]
```diff
@@ -118,32 +118,59 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
...
-        const char **tmpdata;
+    if (n <= 0)
+        return 1;
+
+    /* Sanity checks */
+    if (n > INT_MAX - st->num
+        || (size_t)n > SIZE_MAX / sizeof(char *) - st->num)
+        return 0;
+
+    target = st->num + n * sizeof(char *);
```
Comment (dot-asm) [inline crypto/stack/stack.c]: Why n is multiplied by `sizeof(char*)`?

File: crypto/stack/stack.c [hunk 4]
```diff
@@ -118,32 +118,59 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
     return (NULL);
 }
 
-int OPENSSL_sk_insert(OPENSSL_STACK *st, const void *data, int loc)
+int OPENSSL_sk_reserve(OPENSSL_STACK *st, int n)
 {
-    if (st == NULL || st->num < 0 || st->num == INT_MAX) {
+    size_t target;
+    size_t num_alloc = st->num_alloc;
```
Comment (dot-asm) [inline crypto/stack/stack.c]: This is redundant assignment. It asks for redundant question why. Well, initially it was just question, and then it turned redundant. But I wish it wasn't posed to start with. I mean remove redundant assignment.

File: include/openssl/stack.h (hunk 2 above)
Comment (paulidale) [reply include/openssl/stack.h]: I started allowing an expansion with a size_t argument but decided that was insane because the count is an int.

File: crypto/stack/stack.c (hunk 3 above)
Comment (paulidale) [reply crypto/stack/stack.c]: Gone now.  Left over from the size_t based expansion's earlier code.
Although it was probably wrong then too :(

File: crypto/stack/stack.c (hunk 4 above)
Comment (paulidale) [reply crypto/stack/stack.c]: It's gone now.

File: crypto/stack/stack.c [hunk 5]
```diff
@@ -118,32 +118,57 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
...
+        sz = num_alloc * 2;
 
         /* Overflow checks */
-        if (doub_num_alloc < st->num_alloc)
+        if (sz < num_alloc)
             return 0;
 
         /* Avoid overflow due to multiplication by sizeof(char *) */
-        if (doub_num_alloc > SIZE_MAX / sizeof(char *))
+        if (sz > SIZE_MAX / sizeof(char *))
```
Comment (dot-asm) [inline crypto/stack/stack.c]: Actually this is not the only boundary. It makes no sense allocate more than INT_MAX[+1] elements, because maximum index in int.

On related note one can wonder if doubling is actually right strategy. Well, it gets kind of "schizophrenic" here. Doubling is probably right in start, then it should rather switch to linear, but you are probably unlikely to even need that many entries. And since amount of entries is not large, then checks for SIZE_MAX feels excessive...

Comment (paulidale) [reply crypto/stack/stack.c]: By my reading of the standard, it looks like SIZE_MAX can be less than INT_MAX.
I've limited the size to INT_MAX (pushed separately but will be squashed).

As for the doubling, I wondered about this too.  I don't know the right strategy.  Do we have data as to the size stacks can grow to inside OpenSSL?  It is a public API, so we can't know for sure.

Comment (paulidale) [reply crypto/stack/stack.c]: As for the size increase, the best reference I've found so far is _Why Are Vectors Efficient?_ by Andrew Koenig from the _Journal of Object-Orientated Programming_ volume 11, number 5, published in 1998.  I've not been able to locate the actual article but the result is presented in [this comp.land.c++moderated article](https://groups.google.com/d/msg/comp.lang.c++.moderated/asH_VojWKJw/Lo51JEmLVboJ).

The conclusion is that growth factors greater than:
Φ = 1.618...}{2}&space;%5Capprox&space;1.618)
aren't great.  I've found other discussions that suggest linear growth is never great, so I'm going to try exponential growth with a factor of 1.5.

**Comment by paulidale** [comment]
Growth by 1.5 done.

File: crypto/stack/stack.c (hunk 5 above)
Comment (dot-asm) [reply crypto/stack/stack.c]: > By my reading of the standard, it looks like SIZE_MAX can be less than INT_MAX.

But in all practical cases it's vice versa and what I'm talking about is rather relation between `SIZE_MAX/sizeof(void *)` vs. `INT_MAX` than `SIZE_MAX` and `INT_MAX`.

> I've limited the size to INT_MAX (pushed separately but will be squashed).

Formally speaking the limit should be `(size_t)INT_MAX+1`. Indeed, last permitted index being INT_MAX means that there are INT_MAX+1 elements allocated. Here is another relevant question. Imagine going back to doubling for a moment, and imagine you wanted to reserve INT_MAX-13 elements. With INT_MAX as cap it would fail, because algorithm would search for closest 2^n larger than requested amount but it will fail check for INT_MAX. For wrong reason. Well, one case of course argue that INT_MAX-13 should never happen and we might as well let it slide. But it does make code counter-intuitive. Check for INT_MAX+1 would address this. But now consider proposed 1.5 growth factor... I don't want to calculate problematic values, what I'm rather trying to ask is if `return 0` as outcome of this test is right? Maybe setting setting  to INT_MAX+1 is more adequate? And by "right" and "adequate" I rather mean rather "least counter-intuitive" than "practical"...

**Comment by dot-asm** [comment]
As for 1.5. Do we recognize that `x + x/2` is more exact expression than `(x/2)*3`? I mean do we choose latter even if we recognize former?

File: crypto/stack/stack.c [hunk 6]
```diff
@@ -12,6 +12,34 @@
 #include "internal/numbers.h"
 #include <openssl/stack.h>
 #include <openssl/objects.h>
+#include <errno.h>
+
+/*
+ * The initial number of nodes in the array.
+ */
+static const int min_nodes = 4;
```
Comment (dot-asm) [inline crypto/stack/stack.c]: I was meaning to ask why do you prefer static const over pre-processor macros? I'm not saying it's wrong/inappropriate, just curious over any particular reason. I can imagine it being handy in debugger, because you can examine the value [as opposite to remembering it or consulting likely another file]...

File: crypto/stack/stack.c [hunk 7]
```diff
@@ -12,6 +12,34 @@
...
+ *  1. The denominator must be smaller than the min_nodes value.
+ *  2. The fraction must be greater than unity.
+ *
+ * The first is because the division is performed first and truncating to
+ * zero is bad.  The second is so that the array does grow.  The truncation
+ * from the division happens before the multiplication, hence ratios close
+ * to unity could be problematic for some sizes (e.g. 6/5 with a size of 18).
+ */
+static const int growth_numerator = 3;
+static const int growth_denominator = 2;
```
Comment (dot-asm) [inline crypto/stack/stack.c]: Question here is all [supported] compilers manage to optimize division as shift...

Comment (dot-asm) [reply crypto/stack/stack.c]: I suppose this also answers the remark about `x+x/2`. I mean as long as one wants to keep the options open, one has to accept that `(x/2)*3` is less exact. Well, it's matter of taste, since assertion is that *any* factor less than 1.618 would do...

File: crypto/stack/stack.c (hunk 5 above)
Comment (dot-asm) [reply crypto/stack/stack.c]: It seems that I'm wrong about INT_MAX+1 thing. stack_st.num is amount of elements, not last index, and it's integer. This means that last possible index is INT_MAX-1, not INT_MAX, as I suggested. But the question about whether or not corner cases similar to INT_MAX-13 should be treated as valid (again, not from practical viewpoint, but formal) remains...

File: crypto/stack/stack.c (hunk 7 above)
Comment (richsalz) [reply crypto/stack/stack.c]: Maybe have a define or function to get the next size, rather than two contants.

File: crypto/stack/stack.c (hunk 6 above)
Comment (paulidale) [reply crypto/stack/stack.c]: Yes to debugging.  It also makes the compiler type check (which can be done in a macro but often isn't).  Compilers have optimised these away for ages now so there is no penalty either for performance or space.

I also think it's a bit neater this way.  I generally find expressions a little easier to read when they're mostly in lower case.

So no compelling reason for or against really.

File: crypto/stack/stack.c (hunk 5 above)
Comment (paulidale) [reply crypto/stack/stack.c]: I think I've addressed the INT_MAX-13 concern, I limit growth to INT_MAX now.

I left the check against SIZE_MAX/sizeof(char *).  There is an earlier check for n+st->num against INT_MAX that prevents that being too large and the loop is checking for INT_MAX.  These prevent num_alloc being greater than INT_MAX but it is still possible that size_t cannot handle that many pointers (e.g. when sizeof(int) == sizeof(size_t)).  Thus, I think the SIZE_MAX check is correct.

File: crypto/stack/stack.c (hunk 7 above)
Comment (paulidale) [reply crypto/stack/stack.c]: I was aware that the division by two became a shift, the expansion shouldn't be a common occurrence so the cost of a division would be acceptable -- some compilers will change this to a multiplication (_Division by invariant integers using multiplication_ by Granlund and Montgomery).

I also realised that `(x/2)*3` is low by _one_ for odd `x`.  Correcting this will add complexity to the overflow checking and that's something I was trying to avoid -- it is difficult enough to reason about these checks and an error in them has consequences.  The latter is the main reason I avoided the 8/5 option which is arguably slightly better (being closer to the magic value Φ).

I'm happy to pull that bit of code out into a separate function.  It's only in `OPENSSL_sk_reserve` at the moment.  I wanted to avoid cluttering that function with a long comment about the rationale behind the ratio choice and the concerns this raises.

**Comment by paulidale** [comment]
The expansion size calculation code is separated into its own function.
I've implemented the x+x/2 suggestion of @dot-asm 
I've also fixed the range checking to avoid the off by one low situation that could sometimes occur with the older code.

File: crypto/stack/stack.c (hunk 7 above)
Comment (levitte) [reply crypto/stack/stack.c]: Is the exact rounding a serious reason for worry here?

Comment (dot-asm) [reply crypto/stack/stack.c]: > Is the exact rounding a serious reason for worry here?

No, but I'd say it's appropriate to explicitly discuss it, so that if "but-what-about" question is risen later, there is record of conscious discussion.

File: crypto/stack/stack.c [hunk 8]
```diff
@@ -21,10 +28,43 @@ struct stack_st {
...
+        /* This shouldn't be possible but check it anyway */
+        if (n >= INT_MAX)
+            return 0;
 
-#include <errno.h>
+        /* Expand the size by a factor of 3/2 if it is within range */
+        sz = n < limit ? n + n / 2 : INT_MAX;
+
+        /* Overflow check, which is possible if SIZE_MAX < INT_MAX */
+        if (sz < n)
```
Comment (dot-asm) [inline crypto/stack/stack.c]: Well, I see contradiction here. If you consider SIZE_MAX < INT_MAX as possibility, then you can't calculate limit the way you do. I mean there is implicit cast to narrower type, so which value would *that* yield? If we make such assumption, then limit should be two thirds of *minimum* between SIZE_MAX/sizeof(void *) and INT_MAX. Note that it incidentally works on ILP32 systems... But note that we don't have to make such assumption, or rather we can *choose* not to make one by declaring that ILP32 is the minimum supported. [Well, formally speaking one can still argue that size_t can be narrower than 32 bits, because there is no S in ILP.] Myself I'm fine  with *not* making the assumption.

File: ssl/ssl_cert.c (hunk 1 above)
Comment (dot-asm) [reply ssl/ssl_cert.c]: I'm actually with Rich on this one. In sense it would be more appropriate to check for return value from reserve call, rather than attempt to check individual insertions. Even though there is a chance that individual assertions can work (because another thread would free some memory after reservation attempt?), it's more likely to fail midway.

File: crypto/stack/stack.c [hunk 9]
```diff
@@ -107,43 +147,59 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
...
 
-        /* Overflow checks */
-        if (doub_num_alloc < st->num_alloc)
-            return 0;
+    /* Sanity checks */
+    if (n > INT_MAX - st->num
+        || (size_t)n > SIZE_MAX / sizeof(char *) - st->num)
+        return 0;
+
+    if ((size_t)st->num + n >= st->num_alloc) {
```
Comment (dot-asm) [inline crypto/stack/stack.c]: Even here *if* one make assumption that size_t is narrower than int, then you'd be better of with implicit casts.

Comment (dot-asm) [reply crypto/stack/stack.c]: Well, I bet warning system will complain about signed/unsigned comparison, so implicit cast won't do, right? But in such case it would be formally correct to cast left part to unsigned int and let implicit casts play out between unsigned int and size_t.

Comment (dot-asm) [reply crypto/stack/stack.c]: Again, *if* we make such assumption. Though one still argue that unsigned int cast of the left part is more readable.

Comment (paulidale) [reply crypto/stack/stack.c]: It was a build warning that prompted me to add the cast.
unsigned int seems reasonable and it is guaranteed to work.

File: crypto/stack/stack.c (hunk 8 above)
Comment (paulidale) [reply crypto/stack/stack.c]: Good catch.

Are you suggesting to set the limit to the minimum?
Or that the SIZE_MAX >= INT_MAX assumption is reasonable?  Like you, I'm not aware of a machine that doesn't have this.

How about defining a constant equal to the minimum of SIZE_MAX/sizeof(char *) and INT_MAX and using this instead of both SIZE_MAX and INT_MAX for the limit checking?

Comment (dot-asm) [reply crypto/stack/stack.c]: > Are you suggesting to set the limit to the minimum?
> Or that the SIZE_MAX >= INT_MAX assumption is reasonable?

Well, what I'm saying is that *IF* we make the assumption that SIZE_MAX can be less than INT_MAX, then we wouldn't have another choice. Then I'm suggesting *not* to make the assumption, because it's reasonable in the landscape of supported configurations(\*). Then I'm saying that minimum(\*\*) might be sensible *anyway* for ILP32 account. In other words I kind of suggest both: set to minimum(\*\*), but without making the assumption. I.e. it's not "or", but "and".

BTW, not making the assumption means that this `if (sz < n)` would have to be omitted. Basically making or not making the assumption is also question of readability. As discussed earlier, readability, among other things, is about how many questions pops in your mind when you read the code, and less is better. What I'm hinting at is that more assumptions lead to less readable code and this `if` is the illustration. I mean when you see it you have to ask yourself "why" and seek for answer [which would be "not applicable in this case" for vast majority].

(\*) Not to mention that it's likely to be made all over anyway. It might be appropriate to reflect the fact that we expect SIZE_MAX to be not less than INT_MAX in test/sanitytest.c, so that we would know when it breaks. [And if it does, we would still have an option to disclaim support, as opposite to fixing the whole thing.]

(\*\*) minimum between SIZE_MAX/sizeof(void *) and INT_MAX, not SIZE_MAX and INT_MAX.

Comment (dot-asm) [reply crypto/stack/stack.c]: > set to minimum, but without making the assumption

Just in case, reason for that is kind of readability. We have to check for SIZE_MAX/sizeof(void*) in either case, don't we? And I'd say that it would be more appropriate if all checks are performed in one place, instead of two. I mean now compute_growth checks for INT_MAX, while SIZE_MAX/sizeof(void*) boundary is enforced elsewhere...

File: crypto/stack/stack.c [hunk 10]
```diff
@@ -21,10 +28,43 @@ struct stack_st {
...
+ * when n is even and 1 mod 3 when n is odd.  When n is odd, int_max = 2^n-1
+ * produces limit = 2*floor(int_max/3) which is thus low by one, hence the
+ * correction term that adds one back for this case.
+ */
+static ossl_inline size_t compute_growth(size_t target, size_t current)
+{
+    const size_t limit = (INT_MAX / 3) * 2 + INT_MAX % 3;
+    size_t sz, n = current;
+
+    while (n <= target) {
```
Comment (dot-asm) [inline crypto/stack/stack.c]: BTW, why `<=` and not `<`? Whole idea behind reservation is that you're telling that you're pretty sure that `target` is all you'll ever need. Then why would you insist on 50% overhang?

Comment (dot-asm) [reply crypto/stack/stack.c]: With this in mind one can actually wonder if reservation should even go through this loop. I mean if you're pretty sure that `target` is all you'll *ever* need, then why have *any* overhang? Even if it turns out that it needs extension after all [as result of single insertion], one can increment from the point of initial reservation.

File: crypto/stack/stack.c [hunk 11]
```diff
@@ -21,10 +28,43 @@ struct stack_st {
...
+        /* This shouldn't be possible but check it anyway */
+        if (n >= INT_MAX)
```
Comment (dot-asm) [inline crypto/stack/stack.c]: The commentary is misleading. Let's say `target` was INT_MAX. It should be legitimate value on 64-bit system [just legitimate, not practical]. But note that given `<=` loop condition and sz assignment below, it will be rejected. Once again, this is not about what's practical, but what's *formally* correct.

Comment (paulidale) [reply crypto/stack/stack.c]: This one shouldn't be possible because of the earlier checks that detect this case.
The loop can set `n` as high as INT_MAX and if that's insufficient the request would have been denied earlier.  I don't like the possibility of infinite loops however.

I'll improve the comment.

Comment (dot-asm) [reply crypto/stack/stack.c]: In the loop body you have sz = ... INT_MAX, n = sz, then go to loop beginning, which is while (n <= target). If target is INT_MAX, it enters the loop and returns 0. When it should have returned INT_MAX.

> would have been denied earlier.

Earlier > INT_MAX is denied, not >= INT_MAX.

Comment (paulidale) [reply crypto/stack/stack.c]: I'll fix the `<=` check.  It should be '<'.

File: crypto/stack/stack.c (hunk 7 above)
Comment (paulidale) [reply crypto/stack/stack.c]: The most accurate downward rounding has been done.  The off by one should be gone.

File: crypto/stack/stack.c (hunk 8 above)
Comment (paulidale) [reply crypto/stack/stack.c]: Using a single limit value simplified the range checking quite a bit.
I've moved the limit calculation code down next to the reserve function so the range checks are in one place.

File: crypto/stack/stack.c (hunk 10 above)
Comment (paulidale) [reply crypto/stack/stack.c]: I'll see if I can figure something out so the space isn't wasted.

File: crypto/stack/stack.c [hunk 12]
```diff
@@ -107,43 +112,99 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
...
+ *
+ * The limit to avoid overflow is spot on.  The modulo three correction term
+ * ensures that the limit is the largest number than can be expanded by the
+ * growth factor without exceeding the hard limit.
+ */
+static ossl_inline int compute_growth(int target, int current)
 {
-    if (st == NULL || st->num < 0 || st->num == INT_MAX) {
-        return 0;
+    const int limit = (max_nodes / 3) * 2 + max_nodes % 3;
```
Comment (dot-asm) [inline crypto/stack/stack.c]: Now when max_nodes is "variable" `max_nodes % 3` term appears somewhat controversial. For known values of 0x7fffffff and 0xffffffff/4 it works, but not in general case. "Works" means that `limit + limit/2` doesn't actually exceed the max_nodes. Or in other words *in general case* you can find a max_nodes value for which `limit + limit/2` would exceed max_nodes. So that preceding commentary is actually misleading, because it gives an impression of general case. On the other hand we are not performing `limit + limit/2` operation, but `limit-1 + (limit-1)/2`... So I suppose suggestion is to either omit the term, or clarify commentary by stating that we are scaling up limit-1 [because it's `current < limit` and not `current <= limit`].

File: crypto/stack/stack.c [hunk 13]
```diff
@@ -107,43 +112,99 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
...
+
+    while (current <= target) {
+        /* Check to see if we're at the hard limit */
+        if (current >= max_nodes)
```
Comment (dot-asm) [inline crypto/stack/stack.c]: Once again, in combination with loop condition this prohibits *formally* legitimate case of target == max_nodes.

File: crypto/stack/stack.c [hunk 14]
```diff
@@ -107,43 +112,99 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
...
+    if (n > max_nodes - st->num)
+        return 0;
 
-        /* Avoid overflow due to multiplication by sizeof(char *) */
-        if (doub_num_alloc > SIZE_MAX / sizeof(char *))
+    /* Figure out the new size */
+    num_alloc = st->num + n;
+    if (!exact) {
+        if (num_alloc <= st->num_alloc)
+            return 1;
```
Comment (dot-asm) [inline crypto/stack/stack.c]: Speaking of this condition. Wouldn't it be appropriate to check for equality in exact case? [For reference, my first reaction was to suggest to take out of `if (!exact)`, but then I realized that it would prevent freeing some memory, which might be desirable.]

**Comment by dot-asm** [comment]
While you're at it, you can fix formatting in OPENSSL_sk_deep_copy declaration...

File: crypto/stack/stack.c [hunk 15]
```diff
@@ -107,43 +112,99 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
...
+        return 0;
+
+    st->data = tmpdata;
+    st->num_alloc = num_alloc;
+    return 1;
+}
+
+int OPENSSL_sk_reserve(OPENSSL_STACK *st, int n)
+{
+    if (st == NULL || st->num < 0)
```
Comment (dot-asm) [inline crypto/stack/stack.c]: I'd like to argue in favour of omission of st == NULL check. Here and elsewhere. Thoughts?

File: crypto/stack/stack.c (hunk 12 above)
Comment (paulidale) [reply crypto/stack/stack.c]: Ouch!  2^n mod 3 is always 0 or 1.  The code as written overestimates by 1 when `max_nodes` mod 3 = 2.  Good catch.

File: crypto/stack/stack.c (hunk 14 above)
Comment (paulidale) [reply crypto/stack/stack.c]: Does the exact + equality fall out with the later `num_alloc == st->num` test? a few lines below?

File: crypto/stack/stack.c (hunk 15 above)
Comment (paulidale) [reply crypto/stack/stack.c]: I'm unsure.  Both the `st == NULL` and `st->num < 0` seem redundant which an opaque structure.
This PR isn't really about cleaning the code up, so I left them unchanged.  I don't want to clutter reviews least something is missed in the range code.

There are quite a number of `return (x)` instances in the file which I've also left alone.

File: crypto/stack/stack.c (hunk 12 above)
Comment (dot-asm) [reply crypto/stack/stack.c]: > 2^n mod 3 is always 0 or 1

??? 2^3 % 3 is 2.

File: crypto/stack/stack.c (hunk 14 above)
Comment (dot-asm) [reply crypto/stack/stack.c]: Oh! But no, it doesn't cut it. Because it's different condition, and one can even wonder what's up with that. What does size of array have to do number of initialized elements? I mean I'm struggling with question why look at st->num at all. At least at reservation time it's likely to be 0, isn't it?

File: crypto/stack/stack.c (hunk 15 above)
Comment (dot-asm) [reply crypto/stack/stack.c]: Not doing it here and now is fine. Omitting even st->num < 0... Hmmm... Yeah, I suppose one can omit even that...

File: crypto/stack/stack.c (hunk 14 above)
Comment (dot-asm) [reply crypto/stack/stack.c]: It the context one can wonder if it makes sense to postpone allocation of st->data till first reservation or insertion. On can also wonder if it's appropriate to increase min_nodes. Every allocation has an overhead and question is how does min_modes of 4 compare to this overhead. I mean it's desirable that overhead/payload ratio is low.

Comment (paulidale) [reply crypto/stack/stack.c]: I see the problem now.  Is there such a thing as _code blindness_?
I've fixed this and moved the minimum test up a bit.

As for postponing the allocation of st->data, it is probably a good idea.
Likewise, `min_nodes` of 4 seems low to be but without some statistics it is impossible to tell.
I'd suggest a separate PR for each.

File: crypto/stack/stack.c [hunk 16]
```diff
@@ -107,43 +112,98 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
...
-                                  sizeof(char *) * doub_num_alloc);
-        if (tmpdata == NULL)
+    if (!exact) {
+        if (num_alloc <= st->num_alloc)
+            return 1;
+        num_alloc = compute_growth(num_alloc, st->num_alloc);
+        if (num_alloc == 0)
             return 0;
+    } else if (num_alloc == st->num_alloc)
+        return 1;
```
Comment (dot-asm) [inline crypto/stack/stack.c]: Style nit. If `if` has {}, then `else` has to have pair too.

[Off for today...]

Comment (paulidale) [reply crypto/stack/stack.c]: Fixed.

File: crypto/stack/stack.c (hunk 14 above)
Comment (paulidale) [reply crypto/stack/stack.c]: I've created #4426 to track the suggestions for the future.

File: crypto/stack/stack.c (hunk 12 above)
Comment (paulidale) [reply crypto/stack/stack.c]: 2^n-1 was the intention.

**Comment by paulidale** [comment]
Thanks for persevering on this one and find niggling off by one problems.
Merged.

**Comment by richsalz** [comment]
Perhaps take a look at and close #4379 ?
